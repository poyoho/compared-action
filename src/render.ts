import * as path from 'path'
import * as fs from 'fs'

export interface TimeRecord {
  timing: number
}

export interface DiffRecord extends TimeRecord {
  diff: number
}

export type Records = Record<string, TimeRecord>

const total = (record: Records) =>
  Object.values(record).reduce((sum, info) => (sum += info.timing), 0)

const diffRecord = (o: Records, n: Records): [string, DiffRecord][] => {
  return Object.entries(n).map<[string, DiffRecord]>(([key, val]) => {
    ;(val as DiffRecord).diff = val.timing - (o[key].timing || 0)
    return [key, val as DiffRecord]
  })
}

const formatDiff = (diff: number) => `${diff > 0 ? `+` : `-`}${diff}`

const formatTable = (
  o: Records,
  n: Records,
  fields: string[],
  sortFn: (a: DiffRecord, b: DiffRecord) => number
) => {
  return diffRecord(o, n)
    .sort((a, b) => sortFn(a[1], b[1]))
    .slice(0, 5)
    .map(
      ([key, val]) =>
        `|${key}${formatFields(val as any, fields)}|${val.timing}|${formatDiff(
          val.diff
        )}|`
    )
    .join('\n')
}

const formatHeader = (fields: string[]) => `|${fields.join('|')}|`
const formatLine = (fields: string[]) =>
  `|${fields.map(() => '----').join('|')}|`

const formatFields = (info: Record<string, string>, fields: string[]) => {
  const res = fields.map((field) => info[field]).join('|')
  if (res != '') {
    return `|${res}`
  }
  return res
}

function renderRecord(oRecord: Records, nRecord: Records, fields: string[]) {
  const nTotalServe = total(nRecord)
  const tableHeader = ['file', ...fields, 'timing', 'diff']
  return [
    `total: ${nTotalServe}ms`,
    `total diff: ${formatDiff(nTotalServe - total(oRecord))}ms`,
    `\n<details><summary> Toggle detail... </summary>`,
    '\n### 🗒️ Top 5 (diff)\n',
    formatHeader(tableHeader),
    formatLine(tableHeader),
    formatTable(oRecord, nRecord, fields, (a, b) => a.diff - a.diff),
    `\n</details>`
  ].join('\n')
}

function loadJSONFile(path: string) {
  return JSON.parse(fs.readFileSync(path, { encoding: 'utf-8' }))
}

export function render(
  oldPaths: string[],
  newPaths: string[],
  fields: string[]
): string {
  return [
    '<!--report-->',
    '## 🏆 compress report',
    oldPaths
      .map((oldPath, idx) => ({
        name: path.basename(oldPath).replace('.json', ''),
        o: loadJSONFile(path.resolve(oldPath)) as Records,
        n: loadJSONFile(path.resolve(newPaths[idx])) as Records
      }))
      .map((info) =>
        [
          `\n### ${info.name}\n`,
          renderRecord(info.o, info.n, fields),
          '\n'
        ].join('\n')
      )
      .join('\n')
  ].join('\n')
}
