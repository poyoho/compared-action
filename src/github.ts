import { render } from './render'
import * as core from '@actions/core'
import { context, getOctokit } from '@actions/github'
import type { GitHub } from '@actions/github/lib/utils'

async function comment(github: InstanceType<typeof GitHub>, body: string) {
  const comment = {
    issue_number: context.issue.number,
    owner: context.repo.owner,
    repo: context.repo.repo,
    body
  }
  let commentId: number | null = null
  const comments = (
    await github.rest.issues.listComments({
      ...context.repo,
      issue_number: context.issue.number
    })
  ).data
  for (const c of comments) {
    if (c.user?.type === 'Bot' && c.body?.includes('<!--report-->')) {
      commentId = c.id
      break
    }
  }
  if (commentId) {
    await github.rest.issues.updateComment({
      comment_id: commentId,
      ...comment
    })
  } else {
    await github.rest.issues.createComment(comment)
  }
}

async function cache(github: InstanceType<typeof GitHub>, oldPaths: string[]) {
  // TODO
}

function getInputAsArray(name: string, options?: core.InputOptions): string[] {
  return core
    .getInput(name, options)
    .split('\n')
    .map((s) => s.trim())
    .filter((x) => x !== '')
}

export async function action() {
  const token = core.getInput('token', { required: true })
  const oldPaths = getInputAsArray('old-paths', { required: true })
  const newPaths = getInputAsArray('new-paths', { required: true })
  const fields = getInputAsArray('fields')
  const github = getOctokit(token)

  if (oldPaths.length !== newPaths.length) {
    throw new Error('input old-paths, new-paths should had the same length')
  }

  await comment(github, render(oldPaths, newPaths, fields))
  await cache(github, oldPaths)
}
