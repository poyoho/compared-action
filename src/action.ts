import * as core from '@actions/core'
import * as cache from '@actions/cache'
import { context } from '@actions/github'
import type { GitHub } from '@actions/github/lib/utils'

export function getInputAsInt(
  name: string,
  options?: core.InputOptions
): number | undefined {
  const value = parseInt(core.getInput(name, options))
  if (isNaN(value) || value < 0) {
    return undefined
  }
  return value
}

export function getInputAsArray(
  name: string,
  options?: core.InputOptions
): string[] {
  return core
    .getInput(name, options)
    .split('\n')
    .map((s) => s.trim())
    .filter((x) => x !== '')
}

function logWarning(message: string): void {
  const warningPrefix = '[warning]'
  core.info(`${warningPrefix}${message}`)
}

function isGhes(): boolean {
  const ghUrl = new URL(
    process.env['GITHUB_SERVER_URL'] || 'https://github.com'
  )
  return ghUrl.hostname.toUpperCase() !== 'GITHUB.COM'
}

// Cache token authorized for all events that are tied to a ref
// See GitHub Context https://help.github.com/actions/automating-your-workflow-with-github-actions/contexts-and-expression-syntax-for-github-actions#github-context
function isValidEvent(): boolean {
  return 'GITHUB_REF' in process.env && Boolean(process.env['GITHUB_REF'])
}

function isExactKeyMatch(key: string, cacheKey?: string): boolean {
  return !!(
    cacheKey &&
    cacheKey.localeCompare(key, undefined, {
      sensitivity: 'accent'
    }) === 0
  )
}

function setCacheState(state: string): void {
  core.saveState('COMPARED_CACHE_RESULT', state)
}

function getCacheState(): string | undefined {
  const cacheKey = core.getState('COMPARED_CACHE_RESULT')
  if (cacheKey) {
    core.debug(`Cache state/key: ${cacheKey}`)
    return cacheKey
  }

  return undefined
}

function isCacheFeatureAvailable(): boolean {
  if (!cache.isFeatureAvailable()) {
    if (isGhes()) {
      logWarning(
        'Cache action is only supported on GHES version >= 3.5. If you are on version >=3.5 Please check with GHES admin if Actions cache service is enabled or not.'
      )
    } else {
      logWarning(
        'An internal error has occurred in cache backend. Please check https://www.githubstatus.com/ for any ongoing issue in actions.'
      )
    }
    return false
  }
  return true
}

export async function comment(
  github: InstanceType<typeof GitHub>,
  body: string
) {
  if (!isValidEvent()) {
    logWarning(
      `Event Validation Error: The event type ${process.env.GITHUB_EVENT_NAME} is not supported because it's not tied to a branch or tag ref.`
    )
    return
  }
  const comment = {
    issue_number: context.issue.number,
    owner: context.repo.owner,
    repo: context.repo.repo,
    body
  }
  let commentId: number | null = null
  const comments = (
    await github.rest.issues.listComments({
      ...context.repo,
      issue_number: context.issue.number
    })
  ).data
  for (const c of comments) {
    if (c.user?.type === 'Bot' && c.body?.includes('<!--report-->')) {
      commentId = c.id
      break
    }
  }
  if (commentId) {
    await github.rest.issues.updateComment({
      comment_id: commentId,
      ...comment
    })
  } else {
    await github.rest.issues.createComment(comment)
  }
}

export async function restoreFiles(primaryKey: string, cachePaths: string[]) {
  try {
    if (!isCacheFeatureAvailable()) {
      return
    }

    // Validate inputs, this can cause task failure
    if (!isValidEvent()) {
      logWarning(
        `Event Validation Error: The event type ${process.env.GITHUB_EVENT_NAME} is not supported because it's not tied to a branch or tag ref.`
      )
      return
    }

    try {
      const cacheKey = await cache.restoreCache(cachePaths, primaryKey)
      if (!cacheKey) {
        core.info(`Cache not found for input keys: ${primaryKey}`)
        return
      }

      setCacheState(cacheKey)
      core.info(`Cache restored from key: ${cacheKey}`)
      core.info(`Cache restored from path: ${cachePaths.join(', ')}`)
    } catch (error: unknown) {
      const typedError = error as Error
      if (typedError.name === cache.ValidationError.name) {
        throw error
      } else {
        logWarning(typedError.message)
      }
    }
  } catch (error: unknown) {
    core.setFailed((error as Error).message)
  }
}

export async function cacheFiles(primaryKey: string, cachePaths: string[], opts: {
  force?: boolean,
  uploadChunkSize?: number
}) {
  try {
    if (!isCacheFeatureAvailable()) {
      return
    }

    if (!isValidEvent()) {
      logWarning(
        `Event Validation Error: The event type ${process.env.GITHUB_EVENT_NAME} is not supported because it's not tied to a branch or tag ref.`
      )
      return
    }

    if (!opts?.force) {
      const state = getCacheState()
      if (isExactKeyMatch(primaryKey, state)) {
        core.info(
          `Cache hit occurred on the primary key ${primaryKey}, not saving cache.`
        )
        return
      }
    }

    try {
      await cache.saveCache(cachePaths, primaryKey, {
        uploadChunkSize: opts?.uploadChunkSize
      })
      core.info(`Cache saved with key: ${primaryKey}`)
      core.info(`Cache saved with path: ${cachePaths.join(', ')}`)
    } catch (error: unknown) {
      const typedError = error as Error
      if (typedError.name === cache.ValidationError.name) {
        throw error
      } else if (typedError.name === cache.ReserveCacheError.name) {
        core.info(typedError.message)
      } else {
        logWarning(typedError.message)
      }
    }
  } catch (error: unknown) {
    logWarning((error as Error).message)
  }
}
