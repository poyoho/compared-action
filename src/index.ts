import * as core from '@actions/core'
import { getOctokit } from '@actions/github'
import {
  getInputAsArray,
  comment,
  restoreFiles,
  cacheFiles,
  getInputAsInt
} from './action'
import { render } from './render'

function handleError(err: any): void {
  console.error(err)
  core.setFailed(`Unhandled error: ${err}`)
}

async function action() {
  const token = core.getInput('token', { required: true })
  const oldPaths = getInputAsArray('old-paths', { required: true })
  const newPaths = getInputAsArray('new-paths', { required: true })
  const fields = getInputAsArray('fields')
  const title = core.getInput('title')
  const needCache = core.getBooleanInput('need-cache')
  const uploadChunkSize = getInputAsInt('upload-chunk-size')
  const github = getOctokit(token)

  if (oldPaths.length !== newPaths.length) {
    throw new Error('input old-paths, new-paths should had the same length')
  }

  core.info('input: ')
  core.info(`old-paths: ${oldPaths}`)
  core.info(`new-paths: ${newPaths}`)
  core.info(`fields: ${fields}`)
  core.info(`title: ${title}`)
  core.info(`force-cache: ${needCache}`)
  core.info(`upload-chunk-size: ${uploadChunkSize}`)

  // use the cache overwrite the oldPaths if had the cache
  await restoreFiles(title, oldPaths)
  await comment(github, render(oldPaths, newPaths, fields, title))
  // cache the oldPaths if not had the cache or config force cache
  if (needCache) {
    await cacheFiles(title, oldPaths, { uploadChunkSize })
  }
}

process.on('unhandledRejection', handleError)
action().catch(handleError)
