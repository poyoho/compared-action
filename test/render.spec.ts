import { describe, test, expect, vitest } from 'vitest'
import { render } from '../src/render'

vitest.mock('@actions/core')

describe('render', () => {
  const oldPaths = ['./sample/case1.json', './sample/case2.json']
  const newPaths = ['./sample/new.case1.json', './sample/new.case2.json']
  const title = 'compared action'

  test('table with fields', () => {
    const table = render(oldPaths, newPaths, ['fields'], title)
    expect(table).toMatchSnapshot()
  })

  test('table no fields', () => {
    const table = render(oldPaths, newPaths, [], title)
    expect(table).toMatchSnapshot()
  })
})
