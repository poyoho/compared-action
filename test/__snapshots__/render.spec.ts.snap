// Vitest Snapshot v1

exports[`render > table no fields 1`] = `
"<!--report-->
## 🏆 compared action

### case1

total: 50ms
total diff: --100ms

<details><summary> Toggle detail... </summary>

### 🗒️ Top 5 (diff)

|file|timing|diff|
|----|----|----|
|a|10|-0|
|b|10|--10|
|c|10|--20|
|d|10|--30|
|e|10|--40|

</details>



### case2

total: 50ms
total diff: --100ms

<details><summary> Toggle detail... </summary>

### 🗒️ Top 5 (diff)

|file|timing|diff|
|----|----|----|
|a|10|-0|
|b|10|--10|
|c|10|--20|
|d|10|--30|
|e|10|--40|

</details>

"
`;

exports[`render > table with fields 1`] = `
"<!--report-->
## 🏆 compared action

### case1

total: 50ms
total diff: --100ms

<details><summary> Toggle detail... </summary>

### 🗒️ Top 5 (diff)

|file|fields|timing|diff|
|----|----|----|----|
|a|a|10|-0|
|b|b|10|--10|
|c|c|10|--20|
|d|d|10|--30|
|e|e|10|--40|

</details>



### case2

total: 50ms
total diff: --100ms

<details><summary> Toggle detail... </summary>

### 🗒️ Top 5 (diff)

|file|fields|timing|diff|
|----|----|----|----|
|a|a|10|-0|
|b|b|10|--10|
|c|c|10|--20|
|d|d|10|--30|
|e|e|10|--40|

</details>

"
`;
